src/sidebar/components/test/tag-editor-test.js

```diff
@@ -66,118 +95,166 @@ describe('TagEditor', function() {
+      assert.equal(wrapper.find('Datalist').prop('list').length, 2);
+      wrapper.update();
+      assert.equal(wrapper.find('Datalist').prop('open'), true);
+      wrapper.find('input').instance().value = ''; // clear input
       wrapper
-        .find('datalist option')
-        .at(0)
-        .prop('value'),
-      'tag4'
-    );
+        .find('input')
+        .simulate('input', { inputType: 'deleteContentBackward' });
+      assert.equal(wrapper.find('Datalist').prop('open'), false);
+    });
+
+    it('does not close the suggestions when deleting only part of the <input> value', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag3';
+      typeInput(wrapper);
+      assert.equal(wrapper.find('Datalist').prop('list').length, 2);
+      assert.equal(wrapper.find('Datalist').prop('open'), true);
+      wrapper.find('input').instance().value = 't'; // non-empty input remains
+      wrapper
+        .find('input')
+        .simulate('input', { inputType: 'deleteContentBackward' });
+      assert.equal(wrapper.find('Datalist').prop('open'), true);
+    });
+
+    it('opens the suggestions on focus if <input> is not empty', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag3';
+      assert.equal(wrapper.find('Datalist').prop('open'), false);
+      wrapper.find('input').simulate('focus', {});
+      assert.equal(wrapper.find('Datalist').prop('open'), true);
+    });
+
+    it('does not open the suggestions on focus if <input> is empty', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').simulate('focus', {});
+      assert.equal(wrapper.find('Datalist').prop('open'), false);
+    });
+
+    it('does not open the suggestions on focus if <input> value is only white space', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = ' ';
+      wrapper.find('input').simulate('focus', {});
+      assert.equal(wrapper.find('Datalist').prop('open'), false);
+    });
+
+    it('closes the suggestions when focus is removed from the <input> field', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'non-empty';
+      typeInput(wrapper);
+      assert.equal(wrapper.find('Datalist').prop('open'), true);
+      document.body.dispatchEvent(new Event('focus'));
+      wrapper.update();
+      assert.equal(wrapper.find('Datalist').prop('open'), false);
+    });
+
+    it('does not render duplicate suggestions', () => {
+      // `tag3` supplied in the `tagList` will be a duplicate value relative
+      // with the fakeTagsService.filter result above.
+      const wrapper = createComponent({
+        editMode: true,
+        tagList: ['tag1', 'tag2', 'tag3'],
+      });
+      wrapper.find('input').instance().value = 'non-empty';
+      typeInput(wrapper);
+      assert.equal(wrapper.find('Datalist').prop('list').length, 1);
```

Comment:
This could be compressed into one assert with `assert.deepEqual(wrapper.find('Datalist').prop('list'), ['tag4'])`. The advantage of using `assert.deepEqual` to check both the length and contents of an array or object vs. a series of assets is that you get a more useful error if the check fails.

Reply:
👍